### ISSUE TITLE
Admin | Sync Template v2 flows cannot be Turned On

### ISSUE THREAD

**Issue by RonRohlfs** [microsoft/powerapps-tools#100]
Following CoE Starter Kit - Setup Instructions.docx for the Core Components Solution. At Step 4: Activate the Sync Template Flows Instructions Option 1 we are unable to complete sub-step 6 "Ensure the flows are Turned On". When we try to Turn On any of these flows, we see an error. Example:

Your flow could not turn on: Admin | Sync Template v2 (Apps). Error: Flow client error returned with status code "Forbidden" and details "{"error":{"code":"ApiPolicyOrThrottlingViolation","message":"Flow '' cannot be activated due to either admin API policies or billing throttling. "}}".
 
![COESetupTurnOnFlowIssue](https://user-images.githubusercontent.com/62032705/76349478-50b0cf00-62c7-11ea-9e5e-4009bf856a8b.png)

**Comment by manuelap-msft** [comment]
When you edit the Flow, have you checked that all connections are authenticated and you don't have the red triangle to indicate connections aren't authenticated?

It also looks like you or your Power Platform Admin have set up DLP Policies (Data Loss Prevention Policies) that prevent the combination of connectors in the Flow to be used together. Could you check what policies you have set up and if the connections used by these Flows are in the same business data group (all Power Platform Management Connectors, Office 365 Outlook, Common Data Service, Office 365 Users are the main ones)

**Comment by RonRohlfs** [comment]
Hi Thanks for the help!

I set up an Environment for the Circle of Excellence Starter Kit, and was able to create a DLP Policy, and add the required connectors for the  Admin | Sync Template v2 Flows to the Business data only Data Group. These Flows are now all Turned On and are appear to be running successfully. :-)

I then continued with the CoE Starter Kit - Setup Instructions.docx Step 5: Set up Audit Log Sync. As you know, this uses a custom connector defined by Office-365-Audit-Logs.swagger.json. The installation of the Office 365 Audit Logs custom connector seemed to go smoothly, and the Test succeeded as described in the setup instructions.

The Admin | Sync Audit Logs flow also appears to have installed properly. However, this Flow cannot  be Turned On. It has an Action (GetActivitiesByOperation) that uses the Office 365 Audit Logs custom connector which I assume means this connector needs to be added to the DLP Policy.

I attempted to add this connector using the Add-CustomConnectorToPolicy PowerShell command based on this article: https://alextofan.com/2019/05/15/how-to-add-http-actions-and-custom-connectors-to-a-data-loss-prevention-dlp-policy-in-powerapps-and-flow/.

The results are not encouraging. Another screen capture is attached showing the connector is added with an incorrect display name, and the Admin | Sync Audit Logs flow still cannot be Turned On. Despite this, the Admin | Sync Audit Logs Flow somehow ran once, and the custom connector with the correct display name is in the connections section (see the other new screen capture attachment).

Any ideas on what I did wrong? Do I need to try to back out the connector from the DLP Policy and try to re-add?

Thanks,

Ron

![COESetupAdminSyncAuditLogsFlow](https://user-images.githubusercontent.com/62032705/76468652-b6758780-63a9-11ea-9ac1-ff4f217605a3.png)


![COESetupTurnOnFlowCustomConnectorIssue](https://user-images.githubusercontent.com/62032705/76468325-ecfed280-63a8-11ea-861a-05d1f41833af.png)

**Comment by manuelap-msft** [comment]
Hello,

I'm glad you got one step further and I have added a task to my backlog to add DLP Requirements to the documentation as well, it's not very clear at the moment.

For the Sync Audit Log flow, are you getting the same error when trying to switch it on as you did before?

There is a Canvas App called DLP CUSTOMIZER part of the Core Components solution that allows you to move a custom connector to the business data only group of a DLP Policy without the need to use PowerShell scripts. Could you try and see if that would work for you? 

To use this, open the DLP Customizer app and wait for the custom connectors to load.
type "Office 365 Audit Log" in the search bar and make sure to select (by clicking on it) the one in your CoE Environment, if you have multiple ones selected. In the view, the Environment name is written in bold.
![image](https://user-images.githubusercontent.com/55705669/76471176-dfb60800-63e9-11ea-8078-8ab365c2766d.png)
 
Then, under Policy select the Policy that you have just set up (Center of Excellence Starter Kit Policy) and click Add to Business Data 
![image](https://user-images.githubusercontent.com/55705669/76471229-083e0200-63ea-11ea-95cc-4427324e52ad.png)

**Comment by RonRohlfs** [comment]
Hi,

I did get the same error for the Sync Audit Flow that I saw for the others when i tried to switch it on. That suggested the custom connector needed to be added to the Business data only group.

The DLP Customizer app appears to work as expected, and the icon/tile "Office 365 Audit Logs" now appears on the Data Policy page.

However, there are now two icon/tiles for the custom connector on the Data Policy page, and attempts to Turn On now give an error with "...is referenced more than once"in the message as shown in the new screen captures.

Don't see an option to remove the unwanted custom connector in the DLP Customizer app. I ran the Remove-CustomConnectorFromPolicy PowerShell command which happily reported a 200 response. However, it appears to make no change to the Policy. Both icon/tiles are still there and the Flow cannot be Turned On.

Is there a recommended uninstall/back-out procedure? If I delete the existing Data Policy and create a new one will all the connections recognize it?

Thanks again for all your help.

![COESetupAdminSyncAuditLogsFlowTwoConnectors](https://user-images.githubusercontent.com/62032705/76539273-c9399c00-643d-11ea-9c5a-b5f4f6346355.png)


![COESetupAdminTwoCustomConnectors](https://user-images.githubusercontent.com/62032705/76537722-6e9f4080-643b-11ea-9ef1-ae2d88d6341e.png)

**Comment by manuelap-msft** [comment]
Apologies for the issues you are facing with DLP! 
Would it be possible for you to delete this policy and re-create it, using only the DLP Customizer to add the custom connector to business data only which should make it show up only once?

I don't think there's a need to create a new environment, the issue seems to be just the set up of the DLP policy.

**Comment by RonRohlfs** [comment]
Hi,

Thanks again for all your help! We made great progress.

We deleted the Data Policy and created a new one naming it, "MSFT Solutions Data Policy".  Used the DLP Customizer to add the Office 365 Audit Logs connector. After a few minute delay, the connector appears both in the browser, and in PowerShell (see new screen captures).

The Admin | Sync Audit Logs flow can now be Turned On and run (see COESetupAdminSyncAuditLogsFlow2.png screen capture).  It looks like the COE Starter Kit installation is successful :-)

We would like to impose on you with a few more questions if that's OK.

The flow run failed at the 60th iteration of the Apply to each Log step with code: 412 and message: "A record with matching key values already exists." Do you have any hints on how to remedy that?

The fact that there are exactly 100 Apply to each Log iterations seems suspicious, and looks like a default $top value like what is seen in the SharePoint Get items action. Can you shed any light on that?

There are two CDS connectors available as shown in the COESetupDataPolicyTwoCDSConnectors.png screen capture: Common Data Service, and Common Data Service (current environment). What is the difference? Do we need both?

Finally, is there a roll-back or uninstall process beyond selecting the Solution an the Solutions page and clicking Delete? We tried to delete the COE Starter Kit Solution from the dev tenant. It showed an error (unfortunately can't find the text :-( ), but the Solution appears to be removed.

![COESetupAdminSyncAuditLogsFlow2](https://user-images.githubusercontent.com/62032705/76779720-087f2a00-6769-11ea-8f17-e183f5add12e.png)
t Log

![COESetupDataPolicyTwoCDSConnectors](https://user-images.githubusercontent.com/62032705/76779274-55163580-6768-11ea-9efb-9910c45ea79c.png)
![COESetupDLPCustomizerBrowserResults](https://user-images.githubusercontent.com/62032705/76779275-55aecc00-6768-11ea-963a-f678362bff2a.png)
![COESetupDLPCustomizerPowerShellResults](https://user-images.githubusercontent.com/62032705/76779276-56dff900-6768-11ea-94ae-2ce7a7e75ae2.png)
![COESetupDLPCustomizer](https://user-images.githubusercontent.com/62032705/76779279-56dff900-6768-11ea-9b0a-c2f8df8fb08a.png)

**Comment by manuelap-msft** [comment]
Fantastic, that's good to know! Thank you also for sharing this in such detail with screenshot - this will help in updating the documentation. I'm working on this as we speak, so hopefully the next person won't face the same issues!

For the error, would you be able to open the failed Flow run and navigate to the exact location where the error occured? Is it the apply to each Flow or apply to each Flow Details? Thank you

Solutions can be removed by Deleting them - there shouldn't be any additional steps for you to take!

Regarding the **Common Data Service connector** - appreciate this can cause a bit of confusion. Here's some more detail - currently, both connectors will be required

**Common Data Service**

- Usable from both Power Apps / Automate
- Less feature rich than Common Data Service (current environment)
- No future investment is going into this connector 
- https://docs.microsoft.com/en-us/connectors/commondataservice/

**Common Data Service (current environment)**
- Usable from Power Automate
- Where the future investments are going for Flows that need to interact with CDS
- https://docs.microsoft.com/en-us/connectors/commondataserviceforapps/

**Comment by RonRohlfs** [comment]
Hi,

Thanks for the info on the CDS connectors.

Please feel free to use any screenshots, but please look to see if I missed blacking out any information that shouldn't be shared like tenant Ids or user UPNs.

The error happened in the Admin | Sync Audit Logs run at the (CDS) Update a record action in the Switch Case: LaunchPowerApp within the Apply to each Log loop (4 new screen shots).

Here's the Update a record json:

INPUTS: Body:
{
  "admin_creationtime": "2020-03-13T14:52:05",
  "admin_title": "PowerApps Launch 2020-03-13T14:52:05",
  "admin_appid": "e57db400-ac22-4fea-a46b-0aa013a50cc3",
  "admin_operation": "LaunchPowerApp",
  "admin_userupn": "Lesley.Lynch@ashford.edu",
  "admin_workload": "PowerApps",
  "_ownerid_type": ""
}

OUTPUTS: Body:
{
  "status": 412,
  "message": "--batchresponse_ce02623b-1c49-4545-9c36-b3ac9b9bdb02\r\nContent-Type: application/http\r\nContent-Transfer-Encoding: binary\r\n\r\nHTTP/1.1 412 Precondition Failed\r\nREQ_ID: f781897b-8c9f-465c-a219-bb37b817bb80\r\nContent-Type: application/json; odata.metadata=minimal\r\nOData-Version: 4.0\r\n\r\n{\"error\":{\"code\":\"0x80040237\",\"message\":\"A record with matching key values already exists.\",\"innererror\":{\"message\":\"Cannot insert duplicate key.\",\"type\":\"System.ServiceModel.FaultException`1[[Microsoft.Xrm.Sdk.OrganizationServiceFault, Microsoft.Xrm.Sdk, Version=9.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35]]\",\"stacktrace\":\"   at Microsoft.Crm.Extensibility.OrganizationSdkServiceInternal.Execute(OrganizationRequest request, InvocationContext invocationContext, CallerOriginToken callerOriginToken, WebServiceType serviceType, Boolean checkAdminMode, ExecutionContext executionContext, Dictionary`2 optionalParameters)\\r\\n   at Microsoft.Crm.Extensibility.OData.CrmODataExecutionContext.Upsert(Entity entity)\\r\\n   at Microsoft.Crm.Extensibility.OData.CrmODataExecutionContext.Update(Entity entity, UpdateOption updateOption)\\r\\n   at Microsoft.Crm.Extensibility.OData.CrmODataServiceDataProvider.UpdateEdmEntity(CrmODataExecutionContext context, String edmEntityName, String entityKeyValue, EdmEntityObject entityObject)\\r\\n   at Microsoft.Crm.Extensibility.OData.EntityController.PatchEntityImplementation(String& entityName, String key, EdmEntityObject entityDelta)\\r\\n   at Microsoft.PowerApps.CoreFramework.ActivityLoggerExtensions.Execute[TResult](ILogger logger, EventId eventId, ActivityType activityType, Func`1 func, IEnumerable`1 additionalCustomProperties)\\r\\n   at Microsoft.Xrm.Telemetry.XrmTelemetryExtensions.Execute[TResult](ILogger logger, XrmTelemetryActivityType activityType, Func`1 func)\\r\\n   at lambda_method(Closure , Object , Object[] )\\r\\n   at System.Web.Http.Controllers.ReflectedHttpActionDescriptor.ActionExecutor.<>c__DisplayClass10.<GetExecutor>b__9(Object instance, Object[] methodParameters)\\r\\n   at System.Web.Http.Controllers.ReflectedHttpActionDescriptor.ExecuteAsync(HttpControllerContext controllerContext, IDictionary`2 arguments, CancellationToken cancellationToken)\\r\\n--- End of stack trace from previous location where exception was thrown ---\\r\\n   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()\\r\\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\\r\\n   at System.Web.Http.Controllers.ApiControllerActionInvoker.<InvokeActionAsyncCore>d__0.MoveNext()\\r\\n--- End of stack trace from previous location where exception was thrown ---\\r\\n   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()\\r\\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\\r\\n   at System.Web.Http.Controllers.ActionFilterResult.<ExecuteAsync>d__2.MoveNext()\\r\\n--- End of stack trace from previous location where exception was thrown ---\\r\\n   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()\\r\\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\\r\\n   at System.Web.Http.Dispatcher.HttpControllerDispatcher.<SendAsync>d__1.MoveNext()\"}}}\r\n--batchresponse_ce02623b-1c49-4545-9c36-b3ac9b9bdb02--",
  "error": {
    "message": "--batchresponse_ce02623b-1c49-4545-9c36-b3ac9b9bdb02\r\nContent-Type: application/http\r\nContent-Transfer-Encoding: binary\r\n\r\nHTTP/1.1 412 Precondition Failed\r\nREQ_ID: f781897b-8c9f-465c-a219-bb37b817bb80\r\nContent-Type: application/json; odata.metadata=minimal\r\nOData-Version: 4.0\r\n\r\n{\"error\":{\"code\":\"0x80040237\",\"message\":\"A record with matching key values already exists.\",\"innererror\":{\"message\":\"Cannot insert duplicate key.\",\"type\":\"System.ServiceModel.FaultException`1[[Microsoft.Xrm.Sdk.OrganizationServiceFault, Microsoft.Xrm.Sdk, Version=9.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35]]\",\"stacktrace\":\"   at Microsoft.Crm.Extensibility.OrganizationSdkServiceInternal.Execute(OrganizationRequest request, InvocationContext invocationContext, CallerOriginToken callerOriginToken, WebServiceType serviceType, Boolean checkAdminMode, ExecutionContext executionContext, Dictionary`2 optionalParameters)\\r\\n   at Microsoft.Crm.Extensibility.OData.CrmODataExecutionContext.Upsert(Entity entity)\\r\\n   at Microsoft.Crm.Extensibility.OData.CrmODataExecutionContext.Update(Entity entity, UpdateOption updateOption)\\r\\n   at Microsoft.Crm.Extensibility.OData.CrmODataServiceDataProvider.UpdateEdmEntity(CrmODataExecutionContext context, String edmEntityName, String entityKeyValue, EdmEntityObject entityObject)\\r\\n   at Microsoft.Crm.Extensibility.OData.EntityController.PatchEntityImplementation(String& entityName, String key, EdmEntityObject entityDelta)\\r\\n   at Microsoft.PowerApps.CoreFramework.ActivityLoggerExtensions.Execute[TResult](ILogger logger, EventId eventId, ActivityType activityType, Func`1 func, IEnumerable`1 additionalCustomProperties)\\r\\n   at Microsoft.Xrm.Telemetry.XrmTelemetryExtensions.Execute[TResult](ILogger logger, XrmTelemetryActivityType activityType, Func`1 func)\\r\\n   at lambda_method(Closure , Object , Object[] )\\r\\n   at System.Web.Http.Controllers.ReflectedHttpActionDescriptor.ActionExecutor.<>c__DisplayClass10.<GetExecutor>b__9(Object instance, Object[] methodParameters)\\r\\n   at System.Web.Http.Controllers.ReflectedHttpActionDescriptor.ExecuteAsync(HttpControllerContext controllerContext, IDictionary`2 arguments, CancellationToken cancellationToken)\\r\\n--- End of stack trace from previous location where exception was thrown ---\\r\\n   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()\\r\\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\\r\\n   at System.Web.Http.Controllers.ApiControllerActionInvoker.<InvokeActionAsyncCore>d__0.MoveNext()\\r\\n--- End of stack trace from previous location where exception was thrown ---\\r\\n   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()\\r\\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\\r\\n   at System.Web.Http.Controllers.ActionFilterResult.<ExecuteAsync>d__2.MoveNext()\\r\\n--- End of stack trace from previous location where exception was thrown ---\\r\\n   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()\\r\\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\\r\\n   at System.Web.Http.Dispatcher.HttpControllerDispatcher.<SendAsync>d__1.MoveNext()\"}}}\r\n--batchresponse_ce02623b-1c49-4545-9c36-b3ac9b9bdb02--\r\n"
  },
  "source": "org6aae4afc.crm.dynamics.com",
  "errors": [],
  "debugInfo": "clientRequestId: 995e9607-8e6a-43bc-842e-10f456c11bdc"
}

![AdminSyncAuditLogsFlowError1](https://user-images.githubusercontent.com/62032705/76789055-52700c00-6779-11ea-9e0e-a9f4c1d40d67.png)
![AdminSyncAuditLogsFlowError4](https://user-images.githubusercontent.com/62032705/76789056-53a13900-6779-11ea-8d2f-7668771574bc.png)
![AdminSyncAuditLogsFlowError3](https://user-images.githubusercontent.com/62032705/76789058-53a13900-6779-11ea-810a-b890eea30a7a.png)
![AdminSyncAuditLogsFlowError2](https://user-images.githubusercontent.com/62032705/76789059-5439cf80-6779-11ea-8cb7-aa9f66fc04eb.png)

**Comment by RonRohlfs** [comment]
Hi,

Thank you so much for all your help! 

All the Admin | Sync flows are Turned On and running successfully.

Still unable to complete the Setup. Now stuck on Step 6: Configure the Power BI Dashboard.

However, that's out-of-scope for this issue. We'll keep plugging away at it.

Thanks again!

Ron

**Comment by manuelap-msft** [comment]
That's awesome, thanks for confirming! Please feel free to open a new issue for the Power BI Dashboard issues you are experiencing and I'll do my best to help!
